### DISCUSSION TITLE
支持 padding: auto

### DISCUSSION THREAD

**Discussion by pearmini** [antvis/G2#4947] [Ideas]
# 支持 padding: auto

目前 G2 5.0 的坐标轴会出现溢出的效果，体验非常的不好。https://github.com/antvis/G2/issues/4903

![image](https://user-images.githubusercontent.com/49330279/236608026-9fa0804b-adc8-4e00-86b2-93d107bac1bd.png)

## 出现原因

G2 出于性能的考虑，padding 的值都有默认值，导致溢出。

## 解决办法

- 方案一：label 在固定宽高里面省略（GUI 解决）
- 方案二：padding 根据轴的文本和 title 自动计算大小

第一种思路是在 GUI 里面解决，第二种是需要在 G2 中处理。两种办法都用存在的必要，不冲突：

- 方案二（大部分情况）：希望 ticks 在什么情况都完整展示。
- 方案一（多视图对齐或者交互）：希望 plot 区域的大小固定的情况，比如在 brush 的过程中：label 会变化，但是不希望 plot 区域变化。

## API 设计

```js
const chart = new Chart({
  padding: 'auto',
});
```

## 实现思路

- 目前只处理轴，先不处理图例。
- 水平轴：
  - 连续轴：计算 label 高度和 title 的高度，以及之间的间距得到高度。如果水平方向重叠不做处理，交给轴的 autoHide 处理。
  - 离散轴：计算 label 高度和 title 的高度，以及之间的间距得到高度。如果水平方向重叠，直接旋转 90 度，然后重新计算高度。
- 竖直轴：计算 label 的宽度，titile 的高度，以及之间的间距得到宽度。

## 未来工作

- 把 auto size 的逻辑从 runtime 转移到每个组件本身。
- 优化性能。

## 参考

`padding: auto` 这个 API 和 VegaLite 里面的 `autosize: { type: 'fit' }` 保持一致，目前 G2 策略应该是 `autosize: { type: 'none' }`。

- @antv/g2 4.0
- vega-lite：https://vega.github.io/vega-lite/docs/size.html#autosize

**Comment by hustcc** [comment]
related: https://github.com/antvis/G2/issues/4903

**Comment by hustcc** [comment]
fixed by https://github.com/antvis/G2/pull/5046
